### PR TITLE
basic plumbing to allow setting client tls for kv clusters

### PR DESCRIPTION
**Pull request by pjjw**
with this change we can distribute certs with langley and use them for auth, rather than exposing a bare etcd cluster to the company. can use a read-only cert in the proxy to avoid anyone accidentally trying to write stuff to KV after looking at the collector.

### CODE REVIEW

File: client/etcd/config_test.go [hunk 1]
```diff
@@ -45,6 +45,15 @@ const testConfig = `
...
               - etcd4:2379
+          cert: foo.crt.pem
+          key: foo.key.pem
+        - zone: z3
+          endpoints:
+              - etcd5:2379
+              - etcd6:2379
+          cert: foo.crt.pem
+          key: foo.key.pem
+		  ca: foo_ca.pem
```
Comment (cw9) [inline client/etcd/config_test.go]: nit: fix indentation

File: client/etcd/client_test.go [hunk 2]
```diff
@@ -274,7 +274,7 @@ func testNewETCDFn(t *testing.T) (newClientFn, func()) {
 	ecluster := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
 	ec := ecluster.RandClient()
 
-	newFn := func(endpoints []string) (*clientv3.Client, error) {
+	newFn := func(endpoints []string, cert string, key string) (*clientv3.Client, error) {
```
Comment (cw9) [inline client/etcd/client_test.go]: seems you also need to add "ca string" to the arguments

File: client/etcd/client.go [hunk 3]
```diff
@@ -240,8 +243,32 @@ func (c *csclient) etcdClientGen(zone string) (*clientv3.Client, error) {
 	return cli, nil
 }
 
-func newClient(endpoints []string) (*clientv3.Client, error) {
-	return clientv3.New(clientv3.Config{Endpoints: endpoints})
+func newClient(endpoints []string, certfile string, keyfile string, ca string) (*clientv3.Client, error) {
+	tlscfg := &tls.Config{}
+	if certfile != "" && keyfile != "" {
```
Comment (cw9) [inline client/etcd/client.go]: should we return error when only one of certifile and keyfile is available?

File: client/etcd/client.go [hunk 4]
```diff
@@ -240,8 +243,32 @@ func (c *csclient) etcdClientGen(zone string) (*clientv3.Client, error) {
...
+			InsecureSkipVerify: true,
+			Certificates:       []tls.Certificate{cert},
+		}
+		if ca != "" {
+			caCert, err := ioutil.ReadFile(ca)
+			if err != nil {
+				return nil, err
+			}
+			caPool := x509.NewCertPool()
+			ok := caPool.AppendCertsFromPEM(caCert)
```
Comment (cw9) [inline client/etcd/client.go]: nit: `if ok := caPool.AppendCertsFromPEM(caCert); !ok {...}`

File: client/etcd/client.go (hunk 3 above)
Comment (pjjw) [reply client/etcd/client.go]: i'll just check certfile only for zero-value, then the error should be handled by the failure in tls.LoadX509KeyPair.

**Comment by pjjw** [comment]
you're gonna have to help me figure out why this dang thing isn't passing CI tomorrow.

File: client/etcd/client.go [hunk 5]
```diff
@@ -36,7 +40,6 @@ import (
 	etcdheartbeat "github.com/m3db/m3cluster/services/heartbeat/etcd"
 	"github.com/m3db/m3cluster/services/leader"
 	"github.com/m3db/m3x/instrument"
-	"github.com/m3db/m3x/log"
```
Comment (jeromefroe) [inline client/etcd/client.go]: Seems `goimports` might have removed this package (`github.com/m3db/m3x/log`) in favor or `log`. I think you'll need to switch it back though since everything things will expect the former. (This is one of those magic things VS Code runs for you by default that are usually pretty helpful but can go awry at times.)

**Comment by pjjw** [comment]
lol wow. that's pretty shitty of it!

**Comment by pjjw** [comment]
more though that there's a submodule called '.ci' that doesn't seem to refer to any remote.

**Comment by jeromefroe** [comment]
@pjjw yea that's the CI stuff, need to run `git submodule update --init --recursive` to pull it down after a fresh clone

File: client/etcd/config_test.go [hunk 6]
```diff
@@ -61,8 +70,20 @@ func TestConfig(t *testing.T) {
...
+		ClusterConfig{Zone: "z1", Endpoints: []string{"etcd1:2379", "etcd2:2379"}},
+		ClusterConfig{
+			Zone:      "z2",
+			Endpoints: []string{"etcd3:2379", "etcd4:2379"},
+			Cert:      "foo.crt.pem",
+			Key:       "foo.key.pem",
+		},
+		ClusterConfig{
+			Zone:      "z3",
+			Endpoints: []string{"etcd3:2379", "etcd4:2379"},
```
Comment (jeromefroe) [inline client/etcd/config_test.go]: Should be `etcd5` and `etcd6` respectively.

File: client/etcd/config.go [hunk 7]
```diff
@@ -30,6 +30,9 @@ import (
 type ClusterConfig struct {
 	Zone      string   `yaml:"zone"`
 	Endpoints []string `yaml:"endpoints"`
+	Cert      string   `yaml:"cert",omitempty`
```
Comment (prateek) [inline client/etcd/config.go]: nit: group these fields into a struct and add a suffix to indicate these string refer to a path. E.g. https://github.com/m3db/m3em/blob/master/services/m3em_agent/config/config.go#L67-L72

File: client/etcd/options.go [hunk 8]
```diff
@@ -61,6 +61,13 @@ type Cluster interface {
 
 	Endpoints() []string
 	SetEndpoints(endpoints []string) Cluster
+
+	Cert() string
```
Comment (prateek) [inline client/etcd/options.go]: nit: `CertPath` and so on ..

File: glide.yaml [hunk 9]
```diff
@@ -25,8 +25,8 @@ import:
   version: ^3.0.0 # ie >= 3.0.0, < 4.0.0
 - package: github.com/apache/thrift
   version: 9549b25c77587b29be4e0b5c258221a4ed85d37a
-- package: github.com/uber-go/atomic
-  version: e682c1008ac17bf26d2e4b5ad6cdd08520ed0b22
+- package: go.uber.org/atomic
+  version: ^1
```
Comment (prateek) [inline glide.yaml]: nice

File: client/etcd/client.go [hunk 10]
```diff
@@ -240,8 +243,31 @@ func (c *csclient) etcdClientGen(zone string) (*clientv3.Client, error) {
 	return cli, nil
 }
 
-func newClient(endpoints []string) (*clientv3.Client, error) {
-	return clientv3.New(clientv3.Config{Endpoints: endpoints})
+func newClient(endpoints []string, certfile string, keyfile string, ca string) (*clientv3.Client, error) {
+	tlscfg := &tls.Config{}
+	if certfile != "" {
```
Comment (prateek) [inline client/etcd/client.go]: suggestion(take/leave): group the cert specific code into a function on the sub-type for the config. 

i.e. 
```
type TLSConfiguration struct {
  KeyPath string `yaml...`
  CrtPath ... 
  CACrtPath .. 
} 

func (t TLSConfiguration) NewTLSConfig() (*tls.Config, error) {
  ... 
}
```

File: placement/algo/sharded.go [hunk 11]
```diff
@@ -168,11 +168,7 @@ func (a rackAwarePlacementAlgorithm) ReplaceInstances(
 			return nil, err
 		}
 		load := loadOnInstance(leavingInstance)
-		if load == 0 {
-			result, _, err := addInstanceToPlacement(ph.GeneratePlacement(), leavingInstance, withShards)
-			return result, err
-		}
-		if !a.opts.AllowPartialReplace() {
+		if load != 0 && !a.opts.AllowPartialReplace() {
```
Comment (prateek) [inline placement/algo/sharded.go]: the changes in this file, glide, and `kv/util/...` seem unrelated to TLS changes. Any reason to group them in this PR?

Comment (pjjw) [reply placement/algo/sharded.go]: rebased from master to try and figure out why CI was barfing, i'll undo the rebase

File: client/etcd/client.go (hunk 10 above)
Comment (pjjw) [reply client/etcd/client.go]: I'd agree with you if there were any other logic in the newClient function beforehand, but as is there's nothing, so all we'd do is add another layer of if err != nil boilerplate around the whole thing. Would prefer to leave as-is.

Comment (cw9) [reply client/etcd/client.go]: I'd +1 to Prateek's point, the tls logic is pretty isolated, if we ever need to change the tls logic we just need to add fields to the TLSConfig struct without having to add another field to this function.

File: client/etcd/config.go (hunk 7 above)
Comment (cw9) [reply client/etcd/config.go]: nit: probably better to do `TLS *TLSConfiguration` to indicate it's optional

File: client/etcd/client.go (hunk 10 above)
Comment (pjjw) [reply client/etcd/client.go]: Like I said, there's no logic to separate- before the TLS logic this function was just a return statement. There's nothing else having to do with TLS that etcd cares about- almost all the other fields available in the tls.Config object are constrained by GRPC's restrictions on TLS, and we'd have to make the get/set fields in the tls config object anyhow, unless I'm misunderstanding your suggestion.
